amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+                match event {
+                    PSP22Event::Transfer { from, to, value } => {
+                        self.env().emit_event(Transfer { from, to, value })
+                    }
+                    PSP22Event::Approval {
+                        owner,
+                        spender,
+                        amount,
+                    } => self.env().emit_event(Approval {
+                        owner,
+                        spender,
+                        amount,
+                    }),
+                }
+            }
+        }
+
+        #[inline]
+        fn token_by_address(&self, address: AccountId) -> contract_ref!(PSP22) {
+            address.into()
+        }
+
+        fn update_rates(&mut self) {
+            let current_time = self.env().block_timestamp();
+            for rate in self.pool.token_rates.iter_mut() {
+                rate.update_rate(current_time);
+            }
+        }
+
+        fn fee_to(&self) -> Option<AccountId> {
+            self.pool.fee_receiver
+        }
+
+        /// Scaled rates are rates multiplied by precision. They are assumed to fit in u128.
+        /// If TOKEN_TARGET_DECIMALS is 18 and RATE_DECIMALS is 12, then rates not exceeding ~340282366 should fit.
+        /// That's because if precision <= 10^18 and rate <= 10^12 * 340282366, then rate * precision < 2^128.
+        fn get_scaled_rates(&self) -> Result<Vec<u128>, MathError> {
+            self.pool
+                .token_rates
+                .iter()
+                .zip(self.pool.precisions.iter())
+                .map(|(rate, &precision)| {
+                    rate.get_rate()
+                        .checked_mul(precision)
+                        .ok_or(MathError::MulOverflow(114))
+                })
+                .collect()
+        }
+
+        fn ensure_onwer(&self) -> Result<(), StablePoolError> {
+            ensure!(
+                self.env().caller() == self.owner,
+                StablePoolError::OnlyOwner
+            );
+            Ok(())
+        }
+
+        fn token_id(&self, token: AccountId) -> Result<usize, StablePoolError> {
+            self.pool
+                .tokens
+                .iter()
+                .position(|&id| id == token)
+                .ok_or(StablePoolError::InvalidTokenId(token))
+        }
+
+        /// Checks if tokens are valid and returns the tokens ids
+        fn check_tokens(
+            &self,
+            token_in: AccountId,
+            token_out: AccountId,
+        ) -> Result<(usize, usize), StablePoolError> {
+            //check token ids
+            let token_in_id = self.token_id(token_in)?;
+            let token_out_id = self.token_id(token_out)?;
+            if token_in_id == token_out_id {
+                return Err(StablePoolError::IdenticalTokenId);
+            }
+            Ok((token_in_id, token_out_id))
+        }
+
+        fn mint_admin_fee(&mut self, fee: u128, token_id: usize) -> Result<(), StablePoolError> {
```

Comment:
Is there a reason why the fee is realised this way rather than how it was done in UniV2? There the fee is "[taken](https://github.com/Cardinal-Cryptography/common-amm/blob/main/amm/contracts/pair/lib.rs#L494-L511)" from the input token. No minting of new shares.

Reply:
It's the same here for the regular fee. This is for the admin fee (protocol fee in UniV2 I guess), and it's also the same -- LP tokens are minted.